File: python/python_src/rapidmesh/adapt.py
"""gmsh-style Doerfler-marking adaptive refinement for surface meshes.

This is the MARK -> REFINE half of the adaptive loop (SOLVE and ESTIMATE belong
to the solver): given per-triangle error indicators, Doerfler-mark the bulk and
turn the marked elements into a background size field -- gmsh-style -- which the
mesher's grading gradient-limits and the Ruppert refinement re-triangulates, so
every adapted mesh stays sliver-free with a guaranteed minimum angle.

Typical loop::

    mesh = geom.surface_mesh(maxh=h)
    for _ in range(n_adapt):
        eta = solver.error_indicators(mesh)      # a-posteriori, per triangle
        mesh = refine_dorfler(geom, mesh, eta, theta=0.5, maxh=h)
"""
import numpy as np


def dorfler_mark(eta, theta=0.5):
    """Doerfler (bulk) marking. Returns the indices of the smallest element set
    whose summed SQUARED indicator reaches ``theta`` of the total
    (``theta`` in (0, 1]; 0.5 is the common choice). The squared convention
    treats ``eta`` as an energy-norm error contribution per element."""
    eta = np.abs(np.asarray(eta, dtype=float))
    e2 = eta * eta
    total = float(e2.sum())
    if total <= 0.0 or eta.size == 0:
        return np.empty(0, dtype=int)
    order = np.argsort(e2)[::-1]
    cum = np.cumsum(e2[order])
    k = int(np.searchsorted(cum, theta * total)) + 1
    return np.sort(order[:k])


def _tri_local_h(points, faces):
    """Mean edge length per triangle -- the local element size."""
    p = np.asarray(points, float)
    f = np.asarray(faces)
    e = np.stack([
        np.linalg.norm(p[f[:, 1]] - p[f[:, 0]], axis=1),
        np.linalg.norm(p[f[:, 2]] - p[f[:, 1]], axis=1),
        np.linalg.norm(p[f[:, 0]] - p[f[:, 2]], axis=1),
    ], axis=1)
    return e.mean(axis=1)


def mark_size_field(geom, mesh, eta, *, theta=0.5, factor=2.0, h_min=0.0):
    """Doerfler-mark by ``eta`` and register the marked elements as point size
    sources at ``local_h / factor`` (clamped to ``h_min`` if > 0). Mutates the
    geometry's size field and returns the marked element indices; re-mesh
    afterwards (e.g. ``geom.surface_mesh(...)``) to realise the refinement."""
    p = np.asarray(mesh.points, float)
    f = np.asarray(mesh.faces)
    marked = dorfler_mark(eta, theta)
    if marked.size == 0:
        return marked
    hloc = _tri_local_h(p, f)[marked] / float(factor)
    if h_min > 0.0:
        hloc = np.maximum(hloc, h_min)
    centroids = p[f[marked]].mean(axis=1)
    geom.refine_near_points([tuple(c) for c in centroids], hloc.tolist())
    return marked


def refine_dorfler(geom, mesh, eta, *, theta=0.5, factor=2.0, h_min=0.0, **mesh_kw):
    """One ESTIMATE -> MARK -> REFINE step: Doerfler-mark, build the background
    size field, and return the remeshed surface. ``mesh_kw`` is forwarded to
    :meth:`Geometry.surface_mesh` (e.g. ``maxh``, ``grading``,
    ``target_triangles``). The mesher's grading limits the size gradient and the
    Ruppert refinement keeps the result sliver-free.

    ``eta`` must be a proper a-posteriori estimator -- the per-element error,
    which SHRINKS as an element refines (e.g. ``area * gradient``, ``h**k``).
    Then the marked set stays small and the loop converges. A pointwise field
    that does NOT shrink on refinement (e.g. a bare ``exp(-r**2)``) makes Doerfler
    re-mark the same region every pass, so the refinement never localizes -- it
    "refines everywhere". ``grading`` trades the apron width of the refined zone
    against the minimum angle: steeper grading localizes more tightly but pushes
    some triangles below the 20 deg Ruppert target (still sliver-free)."""
    mark_size_field(geom, mesh, eta, theta=theta, factor=factor, h_min=h_min)
    return geom.surface_mesh(**mesh_kw)

File: crates/rapidmesh-tet/src/adapt.rs
//! gmsh-style Dörfler-marking adaptive refinement for surface meshes: the
//! MARK -> REFINE half of the adaptive loop. Given per-triangle error
//! indicators, Dörfler-mark the bulk and turn the marked elements into point
//! size sources, which the mesher's grading + Ruppert refinement realise
//! sliver-free. SOLVE and ESTIMATE belong to the solver (rapidmom).

use crate::conform::SurfaceMesh;
use rapidmesh_geom::vec3::{len, sub};

/// Dörfler (bulk) marking: the indices of the smallest element set whose summed
/// SQUARED indicator reaches `theta` of the total (`theta` in `(0, 1]`; 0.5 is
/// the common choice). The squared convention treats `eta` as an energy-norm
/// error contribution per element. Returns the marked indices in ascending order.
pub fn dorfler_mark(eta: &[f64], theta: f64) -> Vec<u32> {
    let e2: Vec<f64> = eta.iter().map(|e| e * e).collect();
    let total: f64 = e2.iter().sum();
    if total <= 0.0 || eta.is_empty() {
        return Vec::new();
    }
    // Indices by descending indicator; ties keep descending index order to match
    // a `argsort(e2)[::-1]` (ascending-stable, then reversed) convention.
    let mut order: Vec<u32> = (0..eta.len() as u32).collect();
    order.sort_by(|&a, &b| {
        e2[b as usize]
            .partial_cmp(&e2[a as usize])
            .unwrap()
            .then(b.cmp(&a))
    });
    // Smallest prefix whose cumulative squared indicator reaches `theta * total`.
    let target = theta * total;
    let mut cum = 0.0;
    let mut k = order.len();
    for (i, &idx) in order.iter().enumerate() {
        cum += e2[idx as usize];
        if cum >= target {
            k = i + 1;
            break;
        }
    }
    let mut marked: Vec<u32> = order[..k].to_vec();
    marked.sort_unstable();
    marked
}

impl SurfaceMesh {
    /// Mean edge length per triangle (the local element size), parallel to `faces`.
    pub fn tri_local_h(&self) -> Vec<f64> {
        let p = &self.points;
        self.faces
            .iter()
            .map(|f| {
                let (a, b, c) = (p[f.tri[0]], p[f.tri[1]], p[f.tri[2]]);
                (len(sub(b, a)) + len(sub(c, b)) + len(sub(a, c))) / 3.0
            })
            .collect()
    }

    /// Dörfler-mark by `eta` and turn the marked elements into point size sources:
    /// returns the marked indices and parallel `(centroid, h)` arrays, with
    /// `h = local_h / factor` (clamped to `h_min` when `h_min > 0`). Feed the
    /// `(centroid, h)` pairs into a size field (`refine_near_points`) and remesh.
    pub fn dorfler_size_points(
        &self,
        eta: &[f64],
        theta: f64,
        factor: f64,
        h_min: f64,
    ) -> (Vec<u32>, Vec<[f64; 3]>, Vec<f64>) {
        let marked = dorfler_mark(eta, theta);
        let hloc = self.tri_local_h();
        let p = &self.points;
        let mut centroids = Vec::with_capacity(marked.len());
        let mut hs = Vec::with_capacity(marked.len());
        for &mi in &marked {
            let f = self.faces[mi as usize].tri;
            centroids.push([
                (p[f[0]][0] + p[f[1]][0] + p[f[2]][0]) / 3.0,
                (p[f[0]][1] + p[f[1]][1] + p[f[2]][1]) / 3.0,
                (p[f[0]][2] + p[f[1]][2] + p[f[2]][2]) / 3.0,
            ]);
            let mut h = hloc[mi as usize] / factor;
            if h_min > 0.0 {
                h = h.max(h_min);
            }
            hs.push(h);
        }
        (marked, centroids, hs)
    }
}

#[cfg(test)]
mod tests {
    use super::dorfler_mark;

    #[test]
    fn marks_the_bulk() {
        // One dominant element (10) and three small ones: theta=0.5 of the total
        // squared error (100 + 3 = 103, half = 51.5) is reached by element 0 alone.
        let eta = [10.0, 1.0, 1.0, 1.0];
        assert_eq!(dorfler_mark(&eta, 0.5), vec![0]);
        // theta=1.0 needs every element.
        assert_eq!(dorfler_mark(&eta, 1.0), vec![0, 1, 2, 3]);
        // All-zero indicators mark nothing.
        assert!(dorfler_mark(&[0.0, 0.0], 0.5).is_empty());
    }
}
